File: Assets/Game/Scripts/Moduls/P_Modul_41.lua
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils

---P_Modul_41
---浮块机关
---@class P_Modul_41 : ModulBase
---@field P_Modul_41_Box PhysicsObject
P_Modul_41 = ModulBase:extend()

function P_Modul_41:new()
  ModulBase.super.new(self)
end

function P_Modul_41:Active()
  ModulBase.Active(self)
  self.P_Modul_41_Box:Physicalize()
  self.P_Modul_41_Box.CollisionID = GamePlay.BallSoundManager:GetSoundCollIDByName('WoodOnlyHit')
end
function P_Modul_41:Deactive()
  self.P_Modul_41_Box:UnPhysicalize(true)
  ModulBase.Deactive(self)
end
function P_Modul_41:Reset()
  ObjectStateBackupUtils.RestoreObject(self.gameObject)
end
function P_Modul_41:Backup()
  ObjectStateBackupUtils.BackUpObject(self.gameObject)
end

function CreateClass:P_Modul_41()
  return P_Modul_41()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_41.cs
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  /// <summary>
  /// 浮块机关
  /// </summary>
  public class P_Modul_41 : ModulBase 
  {
    public PhysicsObject P_Modul_41_Box;

    public P_Modul_41() {
      EnableBallRangeChecker = true;
      BallCheckeRange = 60;
    }

    public override void Active()
    {
      base.Active();
      P_Modul_41_Box.CollisionID = GamePlayManager.Instance.BallSoundManager.GetSoundCollIDByName("WoodOnlyHit");
      P_Modul_41_Box.gameObject.SetActive(true);
      P_Modul_41_Box.Physicalize();
    }

    public override void Deactive()
    {
      P_Modul_41_Box.UnPhysicalize(true);
      base.Deactive();
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }

    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    {
      ObjectStateBackupUtils.RestoreObjectAndChilds(gameObject);
    }

    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObjectAndChilds(gameObject);
    }
  }
}